### PR TITLE
Initial ISX TPD Push

### PR DESCRIPTION
**Pull request by mattdatwork**
This adds the initial version of the ISX TPD which uses loose actions to track things.
Also includes initial version of unit tests.

documentation to follow once I figure out what format to put it in.

### CODE REVIEW

File: Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+using UnityEngine;
+using UnityEngine.Experimental.Input;
+using UnityEngine.Experimental.Input.Controls;
+using UnityEngine.XR.PoseProvider;
+using UnityEngine.XR;
+
+namespace UnityEngine.Experimental.Input.XR
+{
+
+    [DefaultExecutionOrder(-30000)]
```
Comment (StayTalm) [inline Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs]: Could be a good design to put this as a constant in the TrackedPoseDriver itself.
E.g. public const int kExecutionOrder = -30000;

That way someone who wants something to run before or after the TrackedPose driver could just do
[DefaultExecutionOrder(TrackedPoseDriver.kExecutionOrder+1)]

File: Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        UpdateType m_UpdateType = UpdateType.UpdateAndBeforeRender;
+        public UpdateType updateType
+        {
+            get { return m_UpdateType; }
+            set { m_UpdateType = value; }
+        }
+
+        [SerializeField]
+        InputAction m_PositionAction;
+        public InputAction positionAction {  get { return m_PositionAction;  } set { m_PositionAction = value; m_PositionBound = false;  BindActions(); } }
```
Comment (StayTalm) [inline Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs]: You enable the new action in BindActions(), but do you need to disable the previously assigned action?

File: Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+            {
+                m_RotationAction.Disable();
+                m_RotationAction.performed -= OnRotationUpdate;
+                m_RotationBound = false;
+            }
+        }
+
+        void OnPositionUpdate(InputAction.CallbackContext context)
+        {
+            if (m_PositionBound)
```
Comment (StayTalm) [inline Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs]: This almost seems like an assert, if you get a callback but position is not bound, something went wrong internally.

File: Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+        }
+
+        protected virtual void Awake()
+        {           
+#if ENABLE_VR
+            if (HasStereoCamera())
+            {
+                XRDevice.DisableAutoXRCameraTracking(GetComponent<Camera>(), true);
+                Camera c = GetComponent<Camera>();
+                Debug.Log(c.fieldOfView);
```
Comment (StayTalm) [inline Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs]: This should be deleted

File: Packages/manifest.json [hunk 5]
```diff
@@ -1,10 +1,14 @@
...
     "com.unity.ads": "2.0.8",
-    "com.unity.analytics": "2.0.16",
-    "com.unity.package-manager-ui": "1.9.11",
+    "com.unity.analytics": "3.0.9",
+    "com.unity.collab-proxy": "1.2.9",
+    "com.unity.package-manager-ui": "2.0.0-preview.7",
     "com.unity.purchasing": "2.0.3",
-    "com.unity.textmeshpro": "1.2.4",
+    "com.unity.textmeshpro": "1.3.0",
+    "com.unity.xr.oculus.standalone": "1.28.0",
```
Comment (StayTalm) [inline Packages/manifest.json]: This got added the moment you went to try it out, but I think it should be removed for the PR.

**Review by StayTalm — CHANGES_REQUESTED** [review]
Looks good could be a bug on switching actions at runtime though (not disabling the overwritten action).  If that's the case be sure to add a test for that too.

File: Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs (hunk 1 above)
Comment (Rene-Damm) [reply Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs]: Would recommend staying away from execution order entirely. Looks like the component wants to process in sync with input so we should simply hook it in there. ATM there's no callback to process stuff after events have been processed, only one happening before. I think we should just add this callback and then the TPD would update as part of input updating.

File: Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+        }
+
+        void OnRotationUpdate(InputAction.CallbackContext context)
+        {
+            if (m_RotationBound)
+            {
+                var quatControl = context.control as QuaternionControl;
+                if (quatControl != null)
+                {
+                    m_CurrentRotation = quatControl.ReadValue();
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs]: Perform `ReadValue` on context, not on control.

File: Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+        }
+
+        void OnPositionUpdate(InputAction.CallbackContext context)
+        {
+            if (m_PositionBound)
+            {
+                var vec3Control = context.control as Vector3Control;
+                if (vec3Control != null)
+                {
+                    m_CurrentPosition = vec3Control.ReadValue();
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs]: `ReadValue` on context.

File: Packages/com.unity.inputsystem/InputSystem/Unity.InputSystem.asmdef [hunk 8]
```diff
@@ -1,6 +1,8 @@
 {
     "name": "Unity.InputSystem",
-    "references": [],
+    "references": [
+    	"UnityEngine.XR.PoseProvider"
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/InputSystem/Unity.InputSystem.asmdef]: What's in this and where does this come from?

File: Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        }
+
+        public void Reset(GameObject go)
+        {
+            go.transform.position = new Vector3(0.0f, 0.0f, 0.0f);
+            go.transform.rotation = new Quaternion(0.0f, 0.0f, 0.0f, 0.0f);
+        }
+
+        [Test]
+        public void TPDUpdateOptionTest()
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs]: No `XXXTest()` tests. All tests should express the functionality that is being established or the behavior that is being verified. `XXXTest` is a bad pattern. You can't know what the test is really about until you've actually read it.

For guidance, take a look at the other tests. Good names are things like `CanUpdateStateOfDeviceWithEvent`. Out of a good name usually follows better content, too. `XXXTest()` tend to lack focus by never establishing what exactly is being verified.

File: Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+            tpd.poseProviderComponent = tpp;
+            return tpp;
+        }
+
+
+        [TearDown]
+        public void TearDown()
+        {
+
+        }
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs]: Remove.

File: Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs [hunk 11]
```diff
@@ -0,0 +1,237 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Experimental.Input;
+using UnityEngine.Experimental.Input.Controls;
+using UnityEngine.XR.PoseProvider;
+using UnityEngine.XR;
+
+namespace UnityEngine.Experimental.Input.XR
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs]: Put in `UnityEngine.Experimental.Input.Plugins.XR`. Move file respectively.

File: Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+                tmp.position = testpos;
+                tmp.rotation = testrot;
+                output = tmp;
+                return true;
+            }
+        }
+
+        internal static TestTrackedPoseDriverWrapper CreateGameObjectWithTPD()
+        {
+            GameObject go = new GameObject();
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs]: Use `var` for all local variables.

File: Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs [hunk 13]
```diff
@@ -0,0 +1,155 @@
+using UnityEngine;
+using UnityEngine.TestTools;
+using NUnit.Framework;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine.SpatialTracking;
+using System.Linq;
+using UnityEngine.XR.PoseProvider;
+
+namespace UnityEngine.Experimental.Input.XR
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs]: Move this file respectively to TrackedPoseDriver.cs. Kill namespace on tests. Just put them in the global namespace.

File: Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+            Assert.That(tpd.positionAction.bindings[0].path, Is.EqualTo("<Gamepad>/buttonSouth"));
+            tpd.positionAction = action2;
+            Assert.That(tpd.positionAction.bindings[0].path, Is.EqualTo("<Gamepad>/buttonNorth"));
+
+            
+            tpd.rotationAction = action1;
+            Assert.That(tpd.rotationAction.bindings[0].path, Is.EqualTo("<Gamepad>/buttonSouth"));
+            tpd.rotationAction = action2;
+            Assert.That(tpd.rotationAction.bindings[0].path, Is.EqualTo("<Gamepad>/buttonNorth"));        
+        }
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs]: This test doesn't test anything meaningful about TrackedPoseDriver. The checks are on functionality sitting in the core input system and covered by tests elsewhere.

There should be a test that triggers an action and then checks that TrackedPostDeriver has set the resulting transform as expected.

**Review by Rene-Damm — CHANGES_REQUESTED** [review]

File: Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs (hunk 1 above)
Comment (mattdatwork) [reply Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs]: which callback should i use here? this sounds _way_ better than hoping re: priority.

File: Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs (hunk 2 above)
Comment (mattdatwork) [reply Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs]: good catch. doing that.

File: Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs (hunk 4 above)
Comment (mattdatwork) [reply Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs]: agreed

File: Packages/manifest.json (hunk 5 above)
Comment (mattdatwork) [reply Packages/manifest.json]: agreed

File: Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs (hunk 6 above)
Comment (mattdatwork) [reply Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs]: ok!

File: Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs (hunk 7 above)
Comment (mattdatwork) [reply Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs]: ok!

File: Packages/com.unity.inputsystem/InputSystem/Unity.InputSystem.asmdef (hunk 8 above)
Comment (mattdatwork) [reply Packages/com.unity.inputsystem/InputSystem/Unity.InputSystem.asmdef]: single monobehaviour (BasePoseProvider) which is shared between the old/new TPD's. useful for arm models / etc.  also pretty handy for testing as i can create a pose provider and send data to the tpd and make sure it does the right thing. can potentially remove if we've got a better option for things like arm models and tracking random data.

File: Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs (hunk 9 above)
Comment (mattdatwork) [reply Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs]: On it.

File: Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs (hunk 10 above)
Comment (mattdatwork) [reply Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs]: Ok

File: Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs (hunk 11 above)
Comment (mattdatwork) [reply Packages/com.unity.inputsystem/InputSystem/Monobehaviors/TrackedPoseDriver.cs]: On it.

File: Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs (hunk 12 above)
Comment (mattdatwork) [reply Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs]: this makes me sad, but doing it.

File: Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs (hunk 13 above)
Comment (mattdatwork) [reply Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs]: Ok

File: Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs (hunk 14 above)
Comment (mattdatwork) [reply Packages/com.unity.inputsystem/Tests/InputSystem/Monobehaviors/TrackedPoseDriverTests.cs]: I was trying to ensure that I correctly allowed changing the binding (as there's code in there which checks if the binding is bound and i want to make sure people dont screw that up). i think the better option is to do as tom suggests and check to ensure i unbind the option before i re-bind it.

File: Packages/manifest.json [hunk 15]
```diff
@@ -1,10 +1,12 @@
...
   "dependencies": {
     "com.unity.ads": "2.0.8",
-    "com.unity.analytics": "2.0.16",
-    "com.unity.package-manager-ui": "1.9.11",
+    "com.unity.analytics": "3.0.9",
+    "com.unity.collab-proxy": "1.2.9",
+    "com.unity.package-manager-ui": "2.0.0-preview.7",
     "com.unity.purchasing": "2.0.3",
-    "com.unity.textmeshpro": "1.2.4",
+    "com.unity.textmeshpro": "1.3.0",
```
Comment (Rene-Damm) [inline Packages/manifest.json]: Please revert changes to this file.

Comment (mattdatwork) [reply Packages/manifest.json]: on it.

Comment (mattdatwork) [reply Packages/manifest.json]: reverted to 492289b7d33dba5269844c73c2c958112463bf07

File: Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+                case InputUpdateType.Dynamic:
+                    OnUpdate();
+                    break;
+                case InputUpdateType.BeforeRender:
+                    OnBeforeRender();
+                    break;                
+            }
+        }
+
+        protected virtual void FixedUpdate()
```
Comment (StayTalm) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs]: Do we really want to update in FixedUpdate?
At the moment it's no different than the Dynamic update data, and can run multiple times per frame, wasting time setting the same data repeatedly.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs [hunk 17]
```diff
@@ -0,0 +1,240 @@
+using System;
+using UnityEngine.XR;
+
+namespace UnityEngine.Experimental.Input.Plugins.XR
+{   
+    [Serializable]
+    [AddComponentMenu("XR/Tracked Pose Driver (New Input System)")]
+    public class TrackedPoseDriver: MonoBehaviour
+    {
+
```
Comment (StayTalm) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs]: Extra Space in here.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+        private bool HasStereoCamera()
+        {
+            var camera = GetComponent<Camera>();
+            return camera != null && camera.stereoEnabled;
+        }
+
+        protected virtual void PerformUpdate()
+        {
+            if (!enabled)
```
Comment (StayTalm) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs]: This shouldn't be necessary.  You use OnEnable/OnDisable to hook up the update loop.  That means it should always be true when hitting this.

File: Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/TrackedPoseDriverTests.cs [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+       
+        public void Reset(GameObject go)
+        {
+            go.transform.position = new Vector3(0.0f, 0.0f, 0.0f);
+            go.transform.rotation = new Quaternion(0.0f, 0.0f, 0.0f, 0.0f);
+        }
+
+        [Test]
+        [Category("TrackedPoseDriver")]
+        public void TPD_UpdateOptions_EffectProcessedData()
```
Comment (StayTalm) [inline Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/TrackedPoseDriverTests.cs]: Affect instead of effect.

**Review by StayTalm — COMMENTED** [review]
Small notes, all just preference or formatting, so it looks good to me.

**Review by StayTalm — APPROVED** [review]

File: Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs (hunk 16 above)
Comment (mattdatwork) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs]: removed

File: Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs (hunk 17 above)
Comment (mattdatwork) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs]: removed

File: Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs (hunk 18 above)
Comment (mattdatwork) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/XR/TrackedPoseDriver.cs]: removed

File: Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/TrackedPoseDriverTests.cs (hunk 19 above)
Comment (mattdatwork) [reply Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/TrackedPoseDriverTests.cs]: changed

**Pull request merged**